### PR TITLE
add hasTransactionHash function in TransactionPool

### PR DESCRIPTION
**Pull request by TrustHenry**
> Issues #345 Change GossipProtocol.hasTransactionHash() to check tx hash in the pool
To solve
First, I added a new function `hasTransactionHash` in Transactionpool. 
This checks if the transaction has a hash in the transaction pool.

### CODE REVIEW

File: source/agora/common/TransactionPool.d [hunk 1]
```diff
@@ -301,6 +301,68 @@ public class TransactionPool
...
+
+    public bool hasTransactionHash (const ref Hash tx) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM tx_pool " ~
+            "WHERE key = ?", tx[]);
+
+        if (!results.empty)
+            return true;
+
+        return false;
```
Comment (AndrejMitrovic) [inline source/agora/common/TransactionPool.d]:
```suggestion
        return !results.empty;
```

File: source/agora/common/TransactionPool.d [hunk 2]
```diff
@@ -301,6 +301,68 @@ public class TransactionPool
...
+            tx = the transaction hash
+
+        Returns:
+            true if the has transaction hash.
+
+    ***************************************************************************/
+
+    public bool hasTransactionHash (const ref Hash tx) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM tx_pool " ~
```
Comment (AndrejMitrovic) [inline source/agora/common/TransactionPool.d]: Technically you do not need the value here, so you could do `SELECT key FROM tx_pool WHERE key = ?`

File: source/agora/common/TransactionPool.d [hunk 3]
```diff
@@ -301,6 +301,68 @@ public class TransactionPool
...
+        assert(pool.hasTransactionHash(hash));
+        pool.remove(tx);
+        assert(!pool.hasTransactionHash(hash));
+    }
+
+    txs.each!(tx => pool.add(tx));
+    assert(pool.length == txs.length);
+    const(Hash) hash = Hash.init;
+    assert(!pool.hasTransactionHash(hash));
+    // 'or 1=1-- SQL Injection Check
```
Comment (AndrejMitrovic) [inline source/agora/common/TransactionPool.d]: Why is this called an "injection check"? I'm not sure how it's trying to inject anything?

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Small changes requested.

File: source/agora/common/TransactionPool.d [hunk 4]
```diff
@@ -301,6 +301,68 @@ public class TransactionPool
...
+
+    /***************************************************************************
+
+        Check if a transaction hash exists in the transaction pool.
+
+        Params:
+            tx = the transaction hash
+
+        Returns:
+            true if the has transaction hash.
```
Comment (AndrejMitrovic) [inline source/agora/common/TransactionPool.d]:
```suggestion
            true if the transaction pool has the transaction hash.
```

File: source/agora/common/TransactionPool.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/common/TransactionPool.d]: Suggested change

File: source/agora/common/TransactionPool.d (hunk 2 above)
Comment (TrustHenry) [reply source/agora/common/TransactionPool.d]: right, Suggested change

File: source/agora/common/TransactionPool.d (hunk 3 above)
Comment (TrustHenry) [reply source/agora/common/TransactionPool.d]: I put it in a test.
We receive parameters through API.
Although converted to hash (bitblob!512), we checked for SQL injection attacks.

Comment (AndrejMitrovic) [reply source/agora/common/TransactionPool.d]: Ah right, I see the comment now. That's a nice idea. I would also (maybe in another pull request) add tests for `pool.add`, making sure injections don't work.

Comment (AndrejMitrovic) [reply source/agora/common/TransactionPool.d]: Ah, never mind. I just realized `add` takes the transaction, not the hash. Ignore my comment.

File: source/agora/common/TransactionPool.d (hunk 4 above)
Comment (TrustHenry) [reply source/agora/common/TransactionPool.d]: Suggested change

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
LGTM. @Geod24?

File: source/agora/common/TransactionPool.d [hunk 5]
```diff
@@ -301,6 +301,65 @@ public class TransactionPool
...
+        pool.remove(tx);
+        assert(!pool.hasTransactionHash(hash));
+    }
+
+    txs.each!(tx => pool.add(tx));
+    assert(pool.length == txs.length);
+    const(Hash) hash = Hash.init;
+    assert(!pool.hasTransactionHash(hash));
+    // 'or 1=1-- SQL Injection attack Check
+    static immutable SqlInjectHash =
```
Comment (Geod24) [inline source/agora/common/TransactionPool.d:353]: Are we looking up the binary data instead of its string representation ???

Comment (AndrejMitrovic) [reply source/agora/common/TransactionPool.d:353]: The key is stored as a binary blob.

Comment (TrustHenry) [reply source/agora/common/TransactionPool.d:353]: That's right now.

File: source/agora/common/TransactionPool.d [hunk 6]
```diff
@@ -301,6 +301,65 @@ public class TransactionPool
...
+    auto pool = new TransactionPool(":memory:");
+    scope(exit) pool.shutdown();  // note: must call outside destructor
+
+    auto gen_key = getGenesisKeyPair();
+    auto txs = makeChainedTransactions(gen_key, null, 1);
+
+    txs.each!(tx => pool.add(tx));
+    assert(pool.length == txs.length);
+
+    foreach (Transaction tx; txs)
```
Comment (MichaelKim20) [inline source/agora/common/TransactionPool.d]:
```suggestion
    foreach (tx; txs)
```
I have not seen the use of type elsewhere. Most of them were omitted.

**Review by MichaelKim20 — COMMENTED** [review]
Good works!

File: source/agora/common/TransactionPool.d (hunk 6 above)
Comment (TrustHenry) [reply source/agora/common/TransactionPool.d]: It's simple.

**Review by MichaelKim20 — APPROVED** [review]

**Review by Geod24 — APPROVED** [review]

**Comment by TrustHenry** [comment]
It's Greeeeeen.

**Pull request merged**